prql/src/translator.rs

```diff
@@ -52,67 +52,49 @@ impl TryFrom<Query> for sql_ast::Query {
             .iter()
             .filter_map(|node| node.item.as_pipeline().cloned())
             .into_only()?;
+        tables.push(pipeline.into());
 
-        sql_query_of_pipeline_and_tables(&pipeline, &tables)
+        sql_query_of_tables(tables)
     }
 }
 
-impl Table {
-    fn to_sql_cte(&self) -> Result<sql_ast::Cte> {
-        let alias = sql_ast::TableAlias {
-            name: Item::Ident(self.name.clone()).try_into()?,
-            columns: vec![],
-        };
-        Ok(sql_ast::Cte {
-            alias,
-            query: sql_query_of_pipeline_and_tables(&self.pipeline, &[])?,
-            from: None,
-        })
+fn sql_query_of_tables(tables: Vec<Table>) -> Result<sql_ast::Query> {
+    // split to atomics
+    let mut atomics = atomic_tables_of_tables(tables)?;
+    if atomics.is_empty() {
+        bail!("No tables?");
     }
-}
 
-fn sql_query_of_pipeline_and_tables(
-    pipeline: &Pipeline,
-    tables: &[Table],
-) -> Result<sql_ast::Query> {
-    let atomic_pipelines = atomic_pipelines_of_pipeline(pipeline)?;
-    // Return early if we have a single atomic pipeline.
-    if atomic_pipelines.len() == 1 && tables.is_empty() {
-        return sql_query_of_atomic_pipeline(&atomic_pipelines.into_only()?);
-    }
-    let tables_from_pipeline = tables_of_pipelines(atomic_pipelines)?;
+    // take last table
+    let main_query = atomics.remove(atomics.len() - 1);
+    let ctes = atomics;
 
-    sql_query_of_tables(&[tables, tables_from_pipeline.as_slice()].concat())
-}
+    // convert each of the CTEs
+    let ctes: Vec<_> = ctes.into_iter().map(table_to_sql_cte).try_collect()?;
 
-fn sql_query_of_tables(tables: &[Table]) -> Result<sql_ast::Query> {
-    let ctes = tables.iter().map(|x| x.to_sql_cte()).try_collect()?;
+    // convert main query
+    let mut main_query = sql_query_of_atomic_pipeline(main_query.pipeline)?;
 
-    Ok(sql_ast::Query {
-        with: Some(sql_ast::With {
+    // attach CTEs
+    if !ctes.is_empty() {
+        main_query.with = Some(sql_ast::With {
             cte_tables: ctes,
             recursive: false,
-        }),
-        order_by: vec![],
-        limit: None,
-        offset: None,
-        fetch: None,
-        body: SetExpr::Select(Box::new(Select {
-            selection: None,
-            distinct: false,
-            top: None,
-            projection: vec![Item::Ident("*".to_string()).try_into()?],
-            from: vec![TableWithJoins {
-                relation: table_factor_of_ident(&tables.last().unwrap().name),
-                joins: vec![],
-            }],
-            group_by: vec![],
-            having: None,
-            lateral_views: vec![],
-            sort_by: vec![],
-            cluster_by: vec![],
-            distribute_by: vec![],
-        })),
+        });
+    }
+
+    Ok(main_query)
+}
+
+fn table_to_sql_cte(table: Table) -> Result<sql_ast::Cte> {
```

Comment:
(I've been trying to use `{output}_of_{input}` as a standard naming convention. Not the most important part of the code though!)